### PR TITLE
Added seq get-unsafe and put-unsafe, corresponding tests

### PR DESCRIPTION
**Pull request by Izaakwltn**
Added unsafe options for seq:get and seq:put, with verbose errors.

### CODE REVIEW

File: library/seq.lisp [hunk 1]
```diff
@@ -1,9 +1,9 @@
 (coalton-library/utils:defstdlib-package #:coalton-library/seq
   (:use
-   #:coalton
-   #:coalton-library/builtin
-   #:coalton-library/functions
-   #:coalton-library/classes)
+    #:coalton
```
Comment (stylewarning) [inline library/seq.lisp:3]: nit: not right indentation

File: library/seq.lisp [hunk 2]
```diff
@@ -92,6 +93,15 @@ than `(size seq)`"
         ((Tuple subtree-idx offset) <- (cst-search guess cst idx))
          (subtree <- (vector:index subtree-idx sts))
          (get subtree (- idx offset))))))
+
+  (declare get-unsafe (Seq :a -> UFix -> :a))
+  (define (get-unsafe seq idx)
+    "Gets the member of `seq` at index `idx`, triggers error if `idx` >= `(size seq)`."
+    (match (get seq idx)
+      ((Some x) x)
+      (_None (error (let ((l (math:1- (size seq))))
```
Comment (stylewarning) [inline library/seq.lisp:102]: `((None) ...)`

File: library/seq.lisp [hunk 3]
```diff
@@ -115,6 +125,16 @@ contains `a`."
            (vector:set! subtree-idx new-subtree newsts)
            (pure (RelaxedNode h fss cst newsts)))))))
 
+  (declare put-unsafe ((Seq :a) -> UFix -> :a -> (Seq :a)))
+  (define (put-unsafe seq idx a)
+    "If `idx` is less than `(size seq)`, Return a new `seq` whose `idx` position
+contains `a`, otherwise errors."
+    (match (put seq idx a)
+      ((Some x) x)
+      (_None (error (let ((l (math:1- (size seq))))
```
Comment (stylewarning) [inline library/seq.lisp:134]: `((None) ...)`

**Review by stylewarning — CHANGES_REQUESTED** [review]

File: library/seq.lisp [hunk 4]
```diff
@@ -92,6 +93,15 @@ than `(size seq)`"
...
          (subtree <- (vector:index subtree-idx sts))
          (get subtree (- idx offset))))))
+
+  (declare get-unsafe (Seq :a -> UFix -> :a))
+  (define (get-unsafe seq idx)
+    "Gets the member of `seq` at index `idx`, triggers error if `idx` >= `(size seq)`."
+    (match (get seq idx)
+      ((Some x) x)
+      ((None) (error (let ((l (math:1- (size seq))))
+                   (lisp String (l idx)
```
Comment (stylewarning) [inline library/seq.lisp:103]: style nit: indentation

File: library/seq.lisp [hunk 5]
```diff
@@ -115,6 +125,16 @@ contains `a`."
...
            (pure (RelaxedNode h fss cst newsts)))))))
 
+  (declare put-unsafe ((Seq :a) -> UFix -> :a -> (Seq :a)))
+  (define (put-unsafe seq idx a)
+    "If `idx` is less than `(size seq)`, Return a new `seq` whose `idx` position
+contains `a`, otherwise errors."
+    (match (put seq idx a)
+      ((Some x) x)
+      ((None) (error (let ((l (math:1- (size seq))))
+                   (lisp String (l idx)
```
Comment (stylewarning) [inline library/seq.lisp:135]: style nit: indentation

**Review by stylewarning — APPROVED** [review]

**Comment by Izaakwltn** [comment]
@macrologist suggested reworking `get-unsafe` and `put-unsafe` as altered copies of `get` and `put` so that they're hopefully faster. 

@stylewarning Do these look like better than the existing PR?
```
  (define (get-unsafe seq idx)
    "Get the member of `seq` at index `idx`, or `None` if `idx` >= `(size seq)`"
    (match seq
      ((LeafArray leaves)
       (vector:index-unsafe idx leaves))
      
      ((RelaxedNode _ fss cst sts)
       (unwrap
         (do
          (let guess = (math:div idx fss))
          ((Tuple subtree-idx offset) <- (cst-search guess cst idx))
           (subtree <- (vector:index subtree-idx sts))
           (get subtree (- idx offset)))))))
```
```
  (define (put-unsafe seq idx a)
    "If `idx` is less than `(size seq)`, Return a new `seq` whose `idx` position
contains `a`."
    (match seq
      ((LeafArray leaves)
       (if (<= (vector:length leaves) idx)
           (error (let ((l (math:1- (size seq))))
                    (lisp String (l idx)
                      (cl:format cl:nil "Invalid index ~a for Seq with max index ~a." idx l))))
           (let ((newleaves (vector:copy leaves)))
             (vector:set! idx a newleaves)
             (LeafArray newleaves))))

      ((RelaxedNode h fss cst sts)
       (unwrap
        (do
         (let guess = (math:div idx fss))
         ((Tuple subtree-idx offset) <- (cst-search guess cst idx))
          (subtree <- (vector:index subtree-idx sts))
          (new-subtree <- (put subtree (- idx offset) a))
          (let ((newsts (vector:copy sts)))
            (vector:set! subtree-idx new-subtree newsts)
            (pure (RelaxedNode h fss cst newsts))))))))
```

**Comment by macrologist** [comment]
These are still just calling `get` and `put` respectively. In your definition for the unsafe get, for example, it's just calling the safe `get` at every step except in the special case that you start with a leaf node. So rather than wrapping the second match case in an `unwrap` - instead get in there and eliminate all the dependence on the `Maybe` monad, using `vector:index-unsafe` whenever you can.

**Comment by stylewarning** [comment]
Closing; can be opened as a new PR with a more sharp-edge approach.

**Pull request closed without merging**